main/lsp/wrapper.h

```diff
@@ -40,32 +53,89 @@ class LSPWrapper final {
+               std::unique_ptr<WorkerPool> workers,
+               std::shared_ptr<spd::sinks::ansicolor_stderr_sink_mt> stderrColorSink,
+               std::shared_ptr<spd::logger> typeErrorsConsole);
+
+    static std::unique_ptr<LSPWrapper> createSingleThreaded(std::unique_ptr<core::GlobalState> gs,
+                                                            std::shared_ptr<options::Options> options,
+                                                            const std::shared_ptr<spdlog::logger> &logger,
+                                                            bool disableFastPath = false);
+    static std::unique_ptr<LSPWrapper>
+    createSingleThreaded(std::string_view rootPath = std::string_view(),
+                         std::shared_ptr<options::Options> options = std::make_shared<options::Options>(),
+                         bool disableFastPath = false);
+
+    /**
+     * Create an LSPWrapper in *multithreaded mode*.
+     * In multithreaded mode, LSP uses individual threads for:
+     * - Reading messages
+     * - Watching file system with Watchman
+     * - Preprocessing messages
+     * - Processing messages
+     * - Typechecking requests
+     * ...and it also uses a configurable number of worker threads dedicated to typechecking.
+     *
+     * Note: LSPWrapper _itself_ is not thread safe, so do not interact with it on multiple threads.
+     */
+    static std::unique_ptr<LSPWrapper>
+    createMultiThreaded(std::function<void(std::unique_ptr<LSPMessage>)> &&processResponse,
+                        std::string_view rootPath = std::string_view(),
+                        std::shared_ptr<options::Options> options = std::make_shared<options::Options>(),
+                        int numWorkerThreads = 1, bool disableFastPath = false);
+
     // N.B.: Sorbet assumes we 'own' this object; keep it alive to avoid memory errors.
-    options::Options opts;
+    const std::shared_ptr<options::Options> opts;
+
+    /** If true, then the LSP wrapper is running LSPLoop on a dedicated thread. */
+    const bool multithreadingEnabled;
 
-    LSPWrapper(std::string_view rootPath = EMPTY_STRING, bool disableFastPath = false);
-    LSPWrapper(options::Options &&options, std::string_view rootPath = EMPTY_STRING, bool disableFastPath = false);
-    LSPWrapper(std::unique_ptr<core::GlobalState> gs, options::Options &&options,
-               const std::shared_ptr<spdlog::logger> &logger, bool disableFastPath);
+    /**
+     * Define so:
+     * 1. We can properly destruct unique_ptr<LSPOutputToVector> (which the default destructor can't delete since we
+     * forward decl it in the header)
+     * 2. We can stop lspThread if running in multithreaded mode.
+     */
     ~LSPWrapper();
 
     const LSPConfiguration &config() const;
 
     /**
      * Send a message to LSP, and returns any responses.
+     * NOTE: In multithreaded mode, the responses to these messages are _not_ sent to `processResponse`!
```

Comment:
rather than having a method that does things differently, can we have two methods with different names?

Reply:
? `send` and `getLSPResponsesFor` do have different names?

Reply:
Talked IRL; going to proceed with a two-class setup where:

* There is no multithreaded implementation of `getLSPResponsesFor`. Tests will use a test function that emulate this functionality.
* We don't pass a lambda to LSPWrapper. Instead, messages will queue internally and LSPWrapper will expose a blocking `read` method.